Image inspector

johnhaddon opened the pull request:
> ImageInspector, next? With support for image metadata for xmas?

I heard @lucienfostier had been good this year, so I had the elves see what they could rustle up in a day. The result is something that shows Format, Channels (with min/max/avg stats) and Metadata in separate tabs, with a selector to choose which view to inspect.

![image](https://github.com/GafferHQ/gaffer/assets/1133871/29fa67e8-2ea6-4641-8c58-d5fa6ef5a056)
![image](https://github.com/GafferHQ/gaffer/assets/1133871/071f46a8-6b6d-45d1-8799-ccd8b4e562b3)
![image](https://github.com/GafferHQ/gaffer/assets/1133871/0bd6e4fb-2727-44eb-8031-5e59c771e184)

There's one fly in the ointment, which is this warning in the terminal :

`WARNING : BackgroundTask : Unable to find ScriptNode for ImageInspectorSettings.view`

This comes from the PlugValueWidget for the view menu, which is showing a plug that belongs to the Editor and not to a ScriptNode. Because it can't find a ScriptNode, it won't participate in cancellation, which is problematic. The elves aren't sure how to fix it yet, but it seemed worth getting some comments on the other aspects of the PR, hence this is a draft for now.

lucienfostier commented:
Wonderful! I'm gonna have a look today.

lucienfostier commented on python/GafferImageUI/ImageInspector.py:
```diff
@@ -0,0 +1,460 @@
+
+	def __imagePlugDirtied( self, plug ) :
+
+		if plug.parent() == self._image and self._imagePlugAffectsPath( plug ) :
```
Do you think it would be worth adding a definition for `_imagePlugAffectsPath` to this base class that raise `NotImplementedError` just like for the `_childNames`?

johnhaddon replied:
Good idea - done in https://github.com/GafferHQ/gaffer/pull/5598/commits/0dc621edb840e79b0b61e31a6e244f0539512593.

lucienfostier commented on python/GafferImageUI/ImageInspector.py:
```diff
@@ -0,0 +1,460 @@
+
+	def propertyNames( self, canceller = None ) :
+
+		return Gaffer.Path.propertyNames( self ) + [ "format:value" ]
```
Is that a way to call the base class? What would be the difference with calling `super( _FormatPath, self ).propertyNames()`

johnhaddon replied:
Yes, and I don't think there is any difference with `super()` in this case. I never really bought into `super` before, since it's more longwinded and I'm used to explicitly specifying the base class in C++ anyway. There might be an argument for adopting it now that Python 3 allows us to omit the arguments, but that's not a topic for this PR.

lucienfostier commented on python/GafferImageUI/ImageInspector.py:
```diff
@@ -0,0 +1,460 @@
+		if name != "format:value" :
+			return Gaffer.Path.property( self, name, canceller )
+
+		if len( self ) != 1 :
```
Could you clarify what case this is catching?

johnhaddon replied:
Paths can be any length, with `len( self ) == 0` denoting the root path, and paths where `len( self ) > 1` being invalid (since they could never be created from `_childNames()`). We only have properties on the valid paths.

lucienfostier commented on python/GafferImageUI/ImageInspector.py:
```diff
@@ -0,0 +1,460 @@
+				origin = self._image.format().getDisplayWindow().min()
+				return f"{origin.x}, {origin.y}"
+			elif self[0] == "Data Window Origin" :
+				origin = self._image.dataWindow().min()
```
I think having the data window max would be useful and canonical.

johnhaddon replied:
Added in https://github.com/GafferHQ/gaffer/pull/5598/commits/22ec8a5c3f62905182341005837d316c765e5162. I also renamed the "Origin" fields to "Min" to better associate them with "Max".

lucienfostier commented on python/GafferImageUI/ImageInspector.py:
```diff
@@ -0,0 +1,460 @@
+		visited = set()
+		for key in sorted( metadata.keys() ) :
+			names = key.split( "/" )
+			if len( names ) <= len( selfNames ) or names[:len(selfNames)] != selfNames :
```
Do you think it would be worth adding a comment to explain this without requiring prior
knowledge about `Gaffer::Path`?

johnhaddon replied:
I think it's reasonable to expect that anyone maintaining this code should know (or will look up) that paths use list syntax to access the names. We can't document fundamentals at every point of usage.

lucienfostier reviewed (commented):
LGTM!
That's a fantastic addition to make Gaffer's set of inspectors well rounded.
I was wondering what would you have in mind for deep image?
Would it be nice to have a few stats like total number of samples or avg samples per pixel?
It would be nice also to have the deep state displayed (i.e tidy)

johnhaddon commented:
Thanks for the review @lucienfostier!

> I was wondering what would you have in mind for deep image?
> Would it be nice to have a few stats like total number of samples or avg samples per pixel?
> It would be nice also to have the deep state displayed (i.e tidy)

In https://github.com/GafferHQ/gaffer/pull/5598/commits/a7662256a9e54b6fec224caa5672d5ae1d1a6868 I've added a "Type" field with values "Deep" or "Flat". We don't have any publicly accessible properties saying whether a deep image is tidy or not, and it's not a concept we've really made user-facing in any way, so I haven't added that.

In https://github.com/GafferHQ/gaffer/pull/5598/commits/f59e7f623e4b1385373cdf9bbbeefc2d6bd8214c I've added min/max/average sample counts per pixel. I agree that a total would be useful, but that's not something I could put together out of the C++ components we have right now (I'm trying to limit the scope of this work to elves-in-a-day levels).

johnhaddon commented:
> There's one fly in the ointment, which is this warning in the terminal

I've pushed  https://github.com/GafferHQ/gaffer/pull/5598/commits/8743e2bb39e7e5edb5137fc801444432b302223c to take care of this. Removing Draft status...

lucienfostier commented on python/GafferUI/Editor.py:
```diff
@@ -49,6 +49,34 @@
+			# cancellation support - see `BackgroundTask.cpp`.
+			## \todo Perhaps we can make this more natural at the point we derive
+			# Editor from Node?
+			self["__scriptNode"] = Gaffer.Plug( flags = Gaffer.Plug.Flags.Default & ~Gaffer.Plug.Flags.Serialisable )
```
Perhaps this could be '__scriptPlug'?

murraystevenson commented on startup/gui/layouts.py:
```diff
@@ -68,8 +69,8 @@
 
-layouts.add( 'Standard', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Vertical, 0.974608, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.699765, ( ( GafferUI.SplitContainer.Orientation.Vertical, 0.479631, ( {'tabs': (GafferUI.Viewer( scriptNode ), GafferSceneUI.UVInspector( scriptNode ), GafferDispatchUI.LocalJobs( scriptNode )), 'curren ...
-layouts.add( 'Standard (multi-monitor)', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Vertical, 0.974608, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.699765, ( {'tabs': (GafferUI.GraphEditor( scriptNode ), GafferSceneUI.LightEditor( scriptNode ), GafferUI.AnimationEditor( scriptNode ), GafferSceneUI.PrimitiveInspector( scriptNode )),  ...
+layouts.add( 'Standard', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Vertical, 0.9746079163554892, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.699764982373678, ( ( GafferUI.SplitContainer.Orientation.Vertical, 0.4796310530361261, ( {'tabs': (GafferUI.Viewer( scriptNode ), GafferSceneUI.UVInspector( scriptNode ), GafferDispatchUI.Loca ...
+layouts.add( 'Standard (multi-monitor)', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Vertical, 0.9746079163554892, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.699764982373678, ( {'tabs': (GafferUI.GraphEditor( scriptNode ), GafferSceneUI.LightEditor( scriptNode ), GafferUI.AnimationEditor( scriptNode ), GafferSceneUI.PrimitiveInspect ...
```
I did wonder a bit about the placement of this, and whether it made more sense to put it in the top right corner with the Scene Inspector. Though it does make sense here as the Hierarchy View is unnecessary when working with images (apart from Cryptomatte manifest previews)...

johnhaddon replied:
Yeah, my initial instinct was to put it with the SceneInspector, but after thinking about it a bit I reckon it's better down with the HierarchyView. Mostly because it's more useful to use the NodeEditor in conjunction with the ImageInspector than to use the HierarchyView with it, but also just to keep some balance in the number of editors in each panel.

murraystevenson reviewed (commented):
Nice job elves! A few minor thoughts inline, but this is looking pretty great for a holiday bonus!

Potential future improvements:

- Drag and drop of inspected values, with improved formatting of dragged channel names and metadata keys.
- Filtering of the channel and metadata views, to help isolate specific channels or metadata keys in images with many.

johnhaddon commented on python/GafferImageUI/ImageInspector.py:
```diff
@@ -0,0 +1,494 @@
+
+			if not self._image.channelNames() :
+				# Heuristic for detecting lack of input image.
+				return "---"
```
Good catch - fixed in https://github.com/GafferHQ/gaffer/pull/5598/commits/37a2e1dd0ca79200df4f24e3000fd05071ef68d5.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -60,37 +60,26 @@
 			self["editScope"] = Gaffer.Plug()
 
-	IECore.registerRunTimeTyped( Settings )
+	IECore.registerRunTimeTyped( Settings, typeName = "GafferUI::LightEditor::Settings" )
```
Indeed it should - fixed in https://github.com/GafferHQ/gaffer/pull/5598/commits/81750fa7e4e1df66dde03275b143f866ac09d254.

johnhaddon commented:
Thanks for the review @murraystevenson. I've pushed a new version which squashes all previous fixups and adds new ones to address you comments - see responses inline.

> Drag and drop of inspected values, with improved formatting of dragged channel names and metadata keys.

I did the drag and drop part of this in https://github.com/GafferHQ/gaffer/pull/5598/commits/09bf6067e1e5ce56f53bd6ee7c0b136d0d71d242 because it was pretty straightforward.

> Filtering of the channel and metadata views, to help isolate specific channels or metadata keys in images with many.

I did look into this, but got bogged down after realising that MatchPatternPathFilter isn't quite good enough and that I didn't want to write yet another alternative filter. I've squirreled my meagre progress away to pick up at some point, when I can unify it with the filters in the SetEditor etc.

murraystevenson reviewed (approved):
Changes look good to me. :+1: 

> I did the drag and drop part of this in https://github.com/GafferHQ/gaffer/commit/09bf6067e1e5ce56f53bd6ee7c0b136d0d71d242 because it was pretty straightforward.

Nice one! Thanks!
